metadata/src/main/java/org/apache/kafka/controller/ScramControlManager.java

```diff
@@ -0,0 +1,307 @@
+        if (error.isFailure()) return error;
+        org.apache.kafka.common.security.scram.internals.ScramMechanism internalMechanism =
+            org.apache.kafka.common.security.scram.internals.ScramMechanism.forMechanismName(mechanism.mechanismName());
+        if (upsertion.iterations() < internalMechanism.minIterations()) {
+            return new ApiError(UNACCEPTABLE_CREDENTIAL, "Too few iterations");
+        } else if (upsertion.iterations() > MAX_ITERATIONS) {
+            return new ApiError(UNACCEPTABLE_CREDENTIAL, "Too many iterations");
+        }
+        return ApiError.NONE;
+    }
+
+    ApiError validateDeletion(ScramCredentialDeletion deletion) {
+        ApiError error = validateScramUsernameAndMechanism(deletion.name(),
+            ScramMechanism.fromType(deletion.mechanism()));
+        if (error.isFailure()) return error;
+        ScramCredentialKey key = new ScramCredentialKey(deletion.name(),
+            ScramMechanism.fromType(deletion.mechanism()));
+        if (!credentials.containsKey(key)) {
+            return new ApiError(RESOURCE_NOT_FOUND,
+                "Attempt to delete a user credential that does not exist");
+        }
+        return ApiError.NONE;
+    }
+
+    static ApiError validateScramUsernameAndMechanism(
+        String username,
+        ScramMechanism mechanism
+    ) {
+        if (username.isEmpty()) {
+            return new ApiError(UNACCEPTABLE_CREDENTIAL, "Username must not be empty");
+        }
+        if (mechanism == ScramMechanism.UNKNOWN) {
+            return new ApiError(UNSUPPORTED_SASL_MECHANISM, "Unknown SCRAM mechanism");
+        }
+        return ApiError.NONE;
+    }
+
+    public void replay(RemoveUserScramCredentialRecord record) {
+        ScramCredentialKey key = new ScramCredentialKey(record.name(),
+            ScramMechanism.fromType(record.mechanism()));
+        if (credentials.remove(key) == null) {
+            throw new RuntimeException("Unable to find credential to delete: " + key);
+        }
+        log.info("Removed SCRAM credential for {} with mechanism {}.",
+            key.username, key.mechanism);
+    }
+
+    public void replay(UserScramCredentialRecord record) {
+        ScramCredentialKey key = new ScramCredentialKey(record.name(),
+            ScramMechanism.fromType(record.mechanism()));
+        ScramCredentialValue value = new ScramCredentialValue(record.salt(),
+            record.saltedPassword(),
+            record.iterations());
+        if (credentials.put(key, value) == null) {
+            log.info("Created new SCRAM credential for {} with mechanism {}.",
+                key.username, key.mechanism);
+        } else {
+            log.info("Modified SCRAM credential for {} with mechanism {}.",
+                key.username, key.mechanism);
+        }
+    }
+
+    ApiMessageAndVersion toRecord(ScramCredentialKey key, ScramCredentialValue value) {
```

Comment:
unused method

Reply:
Keeping this as it may be useful for the bootstrap code.